### PR TITLE
cmd/geth: convert until memory runs out, then batch-write to disk

### PR DESCRIPTION
**Pull request by gballet**
This is the second approach to the MPT -> verkle conversion (alternative to #24815 that checks if the performance is better).

In this approach, the tree is filled up until memory runs out, and then the data is dumped to disk in a batch. This will somewhat avoid trashing, even though thrashing will definitely happen on any system without a large amount of memory.

### CODE REVIEW

File: cmd/geth/snapshot.go [hunk 1]
```diff
@@ -576,3 +601,170 @@ func dumpState(ctx *cli.Context) error {
...
+
+		// Store the basic account data
+		var nonce, balance, version [32]byte
+		binary.LittleEndian.PutUint64(nonce[:8], acc.Nonce)
+		for i, b := range acc.Balance.Bytes() {
+			balance[len(acc.Balance.Bytes())-1-i] = b
+		}
+		// XXX use preimages, accItis the hash of the address
+		versionkey := trieUtils.GetTreeKeyVersion(accIt.Hash().Bytes())
+		vRoot.Insert(versionkey, version[:], convdb.Get)
```
Comment (holiman) [inline cmd/geth/snapshot.go:695]: the `vRoot.Insert` calls are pretty standalone, aren't they? Seems like you could have a goroutine which just reads item (key, value) from a chan, and writes to `vRoot.Insert`. And in this loop you just ship the stuff into the channel.

**Comment by gballet** [comment]
Current run provides the following graph after two hours:

monitor

Flushing a lot of data from disk produces a very small increase in the disk size. There are several explanations for this. One of them is that the in-memory data structure is much less memory-efficient than its on-disk representation. Another one is that this scheme saves a lot of churn because nodes that are quickly updated aren't going as often through the flush cycle, as if they were written each time.

The estimate for the required time is 16 days.

**Comment by gballet** [comment]
After one day, the performance is hardly better: 25 projected days for the conversion

monitor

Note that there is only 16GB of ram on this machine. Doubling the ram should, in theory, make the process more efficient (but not twice as fast). Using DB batches should also make things faster.

**Comment by holiman** [comment]
But you're still doing everything in serial. So while it's writing/reading, it's doing nothing else. And when crunching verkle, that's all it's doing. Splitting it up into different workers, e.g. with channels, should improve it.

**Comment by Ferrydh** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by gballet** [comment]
I am fully aware that this is sequential. The thing is, even sequentially, the other PR runs into some degenerated form of concurrency issue between the DB and the tree, which I'm investigating. Quite happy to find those bugs in a reproducible context before I ride the non-deterministic dragon.

Anyhow, apart from being seemingly bug-free, this method is definitely a dead end: the batch method is just as slow as individually writing values:

monitor

After 39h, the size of the DB and conversion time is exactly the same as without the batch.

**Pull request closed without merging**